File: 3651-minimum-cost-path-with-teleportations/3651-minimum-cost-path-with-teleportations.py
class Solution(object):
    def minCost(self, grid, k):
        m, n = len(grid), len(grid[0])
        max_val = 0
        for row in grid:
            for val in row:
                if val > max_val:
                    max_val = val
                    
        dp = [[float('inf')] * n for _ in range(m)]
        dp[0][0] = 0
        
        for i in range(m):
            for j in range(n):
                if i > 0:
                    dp[i][j] = min(dp[i][j], dp[i-1][j] + grid[i][j])
                if j > 0:
                    dp[i][j] = min(dp[i][j], dp[i][j-1] + grid[i][j])
                    
        for _ in range(k):
            suf = [float('inf')] * (max_val + 2)
            for i in range(m):
                for j in range(n):
                    v = grid[i][j]
                    if dp[i][j] < suf[v]:
                        suf[v] = dp[i][j]
                        
            for v in range(max_val, -1, -1):
                if suf[v+1] < suf[v]:
                    suf[v] = suf[v+1]
                    
            new_dp = [[float('inf')] * n for _ in range(m)]
            for i in range(m):
                for j in range(n):
                    new_dp[i][j] = min(dp[i][j], suf[grid[i][j]])
                    
            for i in range(m):
                for j in range(n):
                    if i > 0:
                        new_dp[i][j] = min(new_dp[i][j], new_dp[i-1][j] + grid[i][j])
                    if j > 0:
                        new_dp[i][j] = min(new_dp[i][j], new_dp[i][j-1] + grid[i][j])
            dp = new_dp
            
        return dp[m-1][n-1]

File: 3651-minimum-cost-path-with-teleportations/3651-minimum-cost-path-with-teleportations.cpp
#include <vector>
#include <algorithm>
#include <map>

class Solution {
public:
    int minCost(std::vector<std::vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        const int INF = 1e9;
        
        std::vector<std::vector<std::vector<int>>> dp(k + 1, std::vector<std::vector<int>>(m, std::vector<int>(n, INF)));
        dp[0][0][0] = 0;
        
        std::map<int, std::vector<std::pair<int, int>>, std::greater<int>> valueToCells;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                valueToCells[grid[i][j]].push_back({i, j});
            }
        }
        
        for (int t = 0; t <= k; ++t) {
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    if (dp[t][i][j] == INF) continue;
                    
                    if (j + 1 < n) {
                        dp[t][i][j + 1] = std::min(dp[t][i][j + 1], dp[t][i][j] + grid[i][j + 1]);
                    }
                    if (i + 1 < m) {
                        dp[t][i + 1][j] = std::min(dp[t][i + 1][j], dp[t][i][j] + grid[i + 1][j]);
                    }
                }
            }
            
            if (t < k) {
                int min_cost_from_higher_src = INF;
                
                for (const auto& group : valueToCells) {
                    int current_group_min = INF;
                    for (const auto& cell : group.second) {
                        current_group_min = std::min(current_group_min, dp[t][cell.first][cell.second]);
                    }
                    
                    min_cost_from_higher_src = std::min(min_cost_from_higher_src, current_group_min);
                    
                    if (min_cost_from_higher_src != INF) {
                        for (const auto& cell : group.second) {
                            dp[t + 1][cell.first][cell.second] = std::min(dp[t + 1][cell.first][cell.second], min_cost_from_higher_src);
                        }
                    }
                }
            }
        }
        
        int ans = INF;
        for (int t = 0; t <= k; ++t) {
            ans = std::min(ans, dp[t][m - 1][n - 1]);
        }
        
        return ans == INF ? -1 : ans;
    }
};